central/auth/datastore/singleton.go

```diff
@@ -15,7 +24,19 @@ var (
 // Singleton provides a singleton auth machine to machine DataStore.
 func Singleton() DataStore {
 	once.Do(func() {
-		ds = New(pgStore.New(globaldb.GetPostgres()))
+		set := m2m.TokenExchangerSetSingleton(roleDataStore.Singleton(), jwt.IssuerFactorySingleton())
+		ds = New(pgStore.New(globaldb.GetPostgres()), set)
+
+		// On initialization of the store, list all existing configs and fill the set.
+		ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS), sac.ResourceScopeKeys(resources.Access)))
+		configs, err := ds.ListAuthM2MConfigs(ctx)
+		utils.Should(err)
+		for _, config := range configs {
+			exchanger, err := set.NewTokenExchangerFromConfig(ctx, config)
+			utils.Should(err)
+			utils.Should(set.UpsertTokenExchanger(exchanger, config.GetId()))
+		}
```

Comment:
[cosmetics] I suggest to move this entire thing to a function with a name so that the intent of the code is more clear.